include/process.h

```diff
@@ -1,23 +1,154 @@
 #ifndef _PROCESS_H_
 #define _PROCESS_H_
 
+#include "atomic.h"
+#include "event.h"
+#include "printf.h"
 #include "vm.h"
 
+#define NR_TASKS (1 << 8)
+
+extern struct PCB* task[NR_TASKS];
+extern int curr_task;
+extern int task_cnt;
+
+enum sig_val {
+    SIGHUP = 1,
+    SIGINT = 2,
+    SIGQUIT = 3,
+    SIGILL = 4,
+    SIGTRAP = 5,
+    SIGABRT = 6,
+    SIGBUS = 7,
+    SIGFPE = 8,
+    SIGKILL = 9,
+    SIGUSR1 = 10,
+    SIGSEGV = 11,
+    SIGUSR2 = 12,
+    SIGPIPE = 13,
+    SIGALRM = 14,
+    SIGTERM = 15,
+    SIGSTKFLT = 16,
+    SIGCHLD = 17,
+    SIGCONT = 18,
+    SIGSTOP = 19,
+    SIGTSTP = 20,
+    SIGTTIN = 21,
+    SIGTTOU = 22,
+    SIGURG = 23,
+    SIGXCPU = 24,
+    SIGXFSZ = 25,
+    SIGVTALRM = 26,
+    SIGPROF = 27,
+    SIGWINCH = 28,
+    SIGIO = 29,
+    SIGPWR = 30,
+    SIGSYS = 31
+};
+
+struct Signal {
+    int val;
+    int from_pid;
+    int status;
+    Signal* next;
+    Signal(int id, int from, int stat) {
+        val = id;
+        from_pid = from;
+        status = stat;
+    }
+};
+
 struct PCB {
     int pid;
     PageTable* page_table;
     SupplementalPageTable* supp_page_table;
+    LockedQueue<Signal, SpinLock>* sigs;
+    Semaphore* waiting_parent;
+    PCB* parent;
+    // fields for maintaining double linked list of children
+    PCB* child_start;
+    PCB* child_end;
+    PCB* next;
+    PCB* before;
 
     PCB() {
-        pid = -1;
+        K::assert(task_cnt < NR_TASKS, "we are out of task space!\n");
+        while (task[curr_task] != nullptr) {
+            curr_task = (curr_task + 1) % NR_TASKS;
+        }
+        task[curr_task] = this;
+        task_cnt += 1;
+        pid = curr_task;
+        page_table = new PageTable;
+        supp_page_table = new SupplementalPageTable;
+        waiting_parent = nullptr;
+        parent = nullptr;
+        child_start = child_end = next = before = nullptr;
+        sigs = new LockedQueue<Signal, SpinLock>;
+    }
+    PCB(int id) {
+        delete task[pid];
```

Comment:
at the start of the constructor, isn't `pid` uninitialized? So accessing`task[pid]` to delete it could be an invalid or strange memory access